package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,266 @@
+            The list has to have the same length as the trajectory.
+            With 'None' the weight of each frame of the trajectory will be the
+            same.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and symmetric,
+            Default: ``None`` sets metric to qcp.CalcRMSDRotationalMatrix
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        """
+        self._u = u
+        self.select = select
+        self.atoms = u.select_atoms(select)
+        self._natoms = self.atoms.n_atoms
+
+        traj = u.trajectory
+        self._epsilon = epsilon
+        if metric is not None:
+            self._metric = metric
+        else:
+            self._metric = qcp.CalcRMSDRotationalMatrix
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+        # modulus to prevent array index out of bounds exception
+        self._k = k % self.nframes
+        if weights is None:
+            # weights do not apply to metric
+            self._weights = np.ones((self.nframes,))
+        else:
+            if weights.shape[0] != self.nframes:
+                raise ValueError("The weight should have the same length as the"
+                                 'trajectory')
+            else:
+                # weights are constructed as relative to the mean
+                self._weights = (np.asarray(weights, dtype=np.float64) /
+                                 np.mean(weights))
+
+    def _prepare(self):
+        self.diffusion_matrix = np.zeros((self.nframes, self.nframes))
+        if self._epsilon == 'average':
+            self._epsilon = np.zeros((self.nframes, ), )
+            self._type_epsilon = 'average'
+        else:
+            value_epsilon = self._epsilon
+            self._epsilon = np.full((self.nframes, ), value_epsilon)
+            self._type_epsilon = 'constant'
+
+        self._rot = np.zeros(9)
+
+    def _single_frame(self):
+        traj_index = self._ts.frame
+        i_ref = self.atoms.positions - self.atoms.center_of_mass()
+
+        # diagonal entries need not be calculated due to metric(x,x) == 0 in
+        # theory, _ts not updated properly. Possible savings by setting a cutoff
+        # for significant decimal places to sparsify matrix
+        for j in range(self.nframes-1, self._ts.frame-1, -1):
+            self._ts = self._u.trajectory[j]
+            j_ref = self.atoms.positions - self.atoms.center_of_mass()
+            ij_result = self._metric(i_ref.T.astype(np.float64),
+                                     j_ref.T.astype(np.float64), self._natoms,
+                                     self._rot, weights=None)
+            if(ij_result < 1E-05):
```

Comment:
That's a good point, thanks
On Fri, Jun 10, 2016 at 7:52 AM kain88-de <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r66626742>:
>
> > +        self._rot = np.zeros(9)
> > +
> > +    def _single_frame(self):
> > +        traj_index = self._ts.frame
> > +        i_ref = self.atoms.positions - self.atoms.center_of_mass()
> > +
> > +        # diagonal entries need not be calculated due to metric(x,x) == 0 in
> > +        # theory, _ts not updated properly. Possible savings by setting a cutoff
> > +        # for significant decimal places to sparsify matrix
> > +        for j in range(self.nframes-1, self._ts.frame-1, -1):
> > +            self._ts = self._u.trajectory[j]
> > +            j_ref = self.atoms.positions - self.atoms.center_of_mass()
> > +            ij_result = self._metric(i_ref.T.astype(np.float64),
> > +                                     j_ref.T.astype(np.float64), self._natoms,
> > +                                     self._rot, weights=None)
> > +            if(ij_result < 1E-05):
>
> you are setting this value very arbitrary. It is below our machine
> precision for one and I might want to have a higher value to force tell the
> diffusion maps when two structures are basically identical.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/de0f209130d7ca01b3ec4a8b987d183584a19d50#r66626742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARplfswLYqN6knI4GO3URUZ7BfUhqks5qKXoogaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs